Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE usage in TextResourceDecoder.cpp

cdumez opened the pull request:
#### 2d4ba2bf5fbe32c832321045184df2e480ae0bc2
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE usage in TextResourceDecoder.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=285655">https://bugs.webkit.org/show_bug.cgi?id=285655</a>

Reviewed by Darin Adler.

* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::find):
(WebCore::KanjiCode::judge):
(WebCore::findXMLEncoding):
(WebCore::TextResourceDecoder::checkForBOM):
(WebCore::TextResourceDecoder::checkForHeadCharset):
(WebCore::bytesEqual): Deleted.

Canonical link: <a href="https://commits.webkit.org/288676@main">https://commits.webkit.org/288676@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79cdfe7283dbcadd44a8632120c0a3cef329fbb9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84100 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3718 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89175 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35108 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86185 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3809 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11686 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65405 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23248 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87146 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2842 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76413 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45699 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2780 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30637 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34157 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31398 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90556 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11365 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8237 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73866 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11589 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73077 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17377 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2712 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13011 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11317 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16789 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11165 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14641 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12937 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/loader/TextResourceDecoder.cpp:
```diff
@@ -36,35 +36,25 @@
+        return notFound;
+
+    size_t sizeDifference = subject.size() - target.size();
+    for (size_t i = 0; i <= sizeDifference; ++i) {
```
Not that it’s changing here, but I’m really surprised that this is `<=` rather than `<`.

darinadler commented on Source/WebCore/loader/TextResourceDecoder.cpp:
```diff
@@ -324,19 +302,15 @@ TextResourceDecoder::~TextResourceDecoder() = default;
-        return true;
-    return data[0] != 0xef || data[1] != 0xbb || data[2] != 0xbf;
+    return !spanHasPrefix(data, std::span { bom });
 }
```
I don’t think we need a function for this any more now that it’s a one-liner.

darinadler commented on Source/WebCore/loader/TextResourceDecoder.cpp:
```diff
@@ -324,19 +302,15 @@ TextResourceDecoder::~TextResourceDecoder() = default;
     auto decoder = TextResourceDecoder::create("text/plain"_s, "UTF-8"_s);
-    if (shouldPrependBOM(data)) {
-        constexpr std::array<uint8_t, 3> bom = { 0xEF, 0xBB, 0xBF };
+    if (shouldPrependBOM(data))
```
```suggestion
    if (!spanHasPrefix(data, std::span { byteOrderMarkUTF8 }))
```

darinadler commented on Source/WebCore/loader/TextResourceDecoder.cpp:
```diff
@@ -36,35 +36,25 @@
-{
-    return *p == b && bytesEqual(p + 1, bs...);
-}
+constexpr std::array<uint8_t, 3> bom = { 0xEF, 0xBB, 0xBF };
```
```suggestion
constexpr std::array<uint8_t, 3> byteOrderMarkUTF8 = { 0xEF, 0xBB, 0xBF };
```
This is a BOM encoded with UTF-8. Also the type of the array is not even `uchar8_t`, the type for UTF-8 characters. So I think the name should include UTF-8 in it.

cdumez committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE usage in TextResourceDecoder.cpp
https://bugs.webkit.org/show_bug.cgi?id=285655

Reviewed by Darin Adler.

* Source/WebCore/loader/TextResourceDecoder.cpp:
(WebCore::find):
(WebCore::KanjiCode::judge):
(WebCore::findXMLEncoding):
(WebCore::TextResourceDecoder::checkForBOM):
(WebCore::TextResourceDecoder::checkForHeadCharset):
(WebCore::bytesEqual): Deleted.

Canonical link: https://commits.webkit.org/288676@main
